py/sentry_streams/runner.py

```diff
@@ -47,7 +53,7 @@ def iterate_edges(p_graph: Pipeline, translator: RuntimeTranslator) -> None:
                     else:
                         next_step: WithInput = cast(WithInput, p_graph.steps[output_step_name])
                         print(f"Apply step: {next_step.name}")
-                        next_step_stream = translator.translate_step(next_step, input_stream)
+                        next_step_stream = translator.translate_step(next_step, input_stream)  # type: ignore
```

Comment:
The type: ignore here is unfortunate and I don't know how to get around it very easily just yet. 

The issue is, the `RuntimeTranslator` has the generic types `Stream, StreamSink`. And the `translate_step` function only works with `Optional[Stream]`. While the code never actually does `translate_step()` on a `StreamSink`, the type checker will complain about this. 

One way around it would be to support a generic `Stream` type that somehow encapsulates both Stream/Sink in the `StreamAdapter`. Thoughts?

Reply:
Not sure I fully get this issue.
Is mypy complaining that `input_stream` is not a `Stream` ?
The fact that you get here with `input_stream` that could be `Stream` or `StreamSink` and rely on the fact that `translate_step` is never called on `StreamSink` points to something not sound on the types we sed around and maybe we should revisit this algorithm entirely. 

So first hint: remove the `type: ignore` and replace it with a `cast`, so the code declares exactly which type is problematic without disabling type checking entirely for the method. 

What happens if you type `step_streams = {}` and allow that to only contain Stream and StreamSource? Where is the type checking failing ?